### DISCUSSION TITLE
Expand the SV value range of Inherited Timing Points

### DISCUSSION THREAD

**Discussion by Kominaru** [ppy/osu#15638] [Ideas]
I know #15016 and #14835 already touch on this issue, but it feels like they didn't move forward or completely lay down the issues with the current lazer and stable behaviour.

As far as I know lazer currently allows 0.1x-10x for inherited timing point SVs, and in stable 0.01x-10x. Talking mainly about Mania, which uses extensively these features to create gimmicky SV maps, these value ranges (both lazer and stable) are far from being enough. In most cases, achieving certain effects is not possible with inherited timing points; this forces mappers to resort to the usage of uninherited timing points (red lines), which ideally should only be needed for timing songs and not for visual effects, specially in the ranked scene. These effects are not really "glitchy", but rather something that should be intuitively achievable with inherited timing points. Moreover, specifically for Mania, the BNG considers these should be rankable in most cases, something not possible if they require red lines to work.

Some of these effects include: 

- **0BPM Red Line gimmicks** (Would be equivalent to a 0xSV green line). Used for stops, stutters, teleports, etc. 
- **10x - ∞x SV** : Used for sightreadable/non-sightreadable teleports. Probably 100000x would be enough o achieve the "instant teleport" effect.
- **Forced Barlines**: Equivalent to one or more uninherited timing point used with to create new barline(s). This is used for visual barline art (used to represent the music) or for barline-assisted SVs (used together with 0BPMs to make them easier to read which boosts enjoyability and vissual appeal; also called "telegraphed SVs"):

Here are some examples that resort to red lines to create these effects and would massively benefit from a more broad behaviour of green lines, as all are renowned tournament SV maps that would become rankable:

- https://osu.ppy.sh/b/1959674 (0BPM gimmicks + "telegraphed" SVs using barlines)
- https://osu.ppy.sh/b/1343180 (0BPM gimmicks)
- https://osu.ppy.sh/b/1306252 (readable teleports, 0BPM gimmicks)
- https://osu.ppy.sh/b/1966772 (0BPM gimmicks)
- https://osu.ppy.sh/b/2158256 (visual barlines for song representation)
- https://osu.ppy.sh/b/1788048 (0BPM gimmicks + "telegraphed" SVs using barlines)

In hindisght, from a non-dev pov, the change doesn't seem extremely large (something akin to the previous fixes to SVs this year at, or the editor snap support for 1/5,1/7, etc.), so we wanted to survey if along with supporting this in lazer, would it be possible to tackle the issue on **stable** as well, since it'd be a massive QoL change for mappers. Forced barlines are probably out of reach in that sense, but it feels like ~0xSV and >10x SV should be doable.

**Comment by peppy** [comment]
Did you open the lazer editor before duplicating multiple existing issues? 0.01x scroll speed adjustments are supported out of the box.

You mention increasing the upper range, but so you also see 0.01 as not being low enough?

**Reply by Kominaru** [reply]
Yes, I have seen 0.01x SVs on action.

 No, it is not enough and all mappers that work with 0BPM SVs agree on this. 

Also I'm opening this issue precisely because you suggested me to do it half a day ago.

**Reply by peppy** [reply]
In your post you said lazer was limited to 0.1, was that a mistake?

How low would SV need to go to be usable? Would it make more sense to leave the limit at 0.01 and add a "freeze" marker instead? That is my current intended direction.

**Reply by Kominaru** [reply]
That was indeed a typo, my apologies. Anything that replicates the behavior of the current 0BPM is likely enough for the purposes of the gimmicks needed. Dont think sub 0.1 SV are used much aside for 0x SV (0BPM), at least in Mania. Only in very specific cases which generally only need to go down to ~0.01x.

For the "forced barlines" part, which is relatively crucial as it affects some SV's playability directly, something as adding artificial "guiding barlines" would suffice as well.

Also just to be sure, I'm guessing that what we're discussing right now is lazer-exclusive, but I want to reiterate for anyone that may be seeing this that stable is also in a quite short-term need of a support for <0.01x and >10x SV (for mapping QoL, evolution of the ranked ecosystem and ease of later map portability to lazer)

**Reply by peppy** [reply]
The reason we aren't adding larger ranges for SV is that it is already overloaded in functionality, and increasing the range is going to *introduce new issues*, especially floating point precision and cross-compatibility issues with lazer. We are already deep enough in these that I think all parties involved in development would agree we do not want this.

This is why I am pushing forward with lazer-first changes for these new behaviours. To make sure we don't shoot ourselves in the foot. Whether backwards compatibility support will be offered or not is dependant on how easy that is to achieve, but the first step is to get things in a good state in lazer.

From what I've understood, these three additions would be enough to resolve your issues:

- The ability to add barlines at any point in time, without requiring a new timing control point.
- The ability to freeze scrolling for a period of time, without requiring 0 BPM or 0x scroll speed.
- The ability to immediately update scrolling to a specific point in time, without requiring infinite BPM or scroll speed.

The benefits of using explicit markers for these behaviours is not limited at making it easy to handle at a code level. It also means we can better warn the user of these changes in a visual way, should the time come that we want to do that (I believe this will be a thing in the future, similar to warning about slider velocity changes in the osu! ruleset).

**Reply by Kominaru** [reply]
Yes, these are the core functionalities needed.  This was the answer we were looking for, I'm eager to seeing whether this is doable enough to offer backwards compatibility in the near future.  Thanks!
